Adds round-start dead bodies to the morgue

zxaber opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Adds two random dead bodies to the morgue room, in two random trays. This is done through a new map helper.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
The dead bodies serve as spare parts supply for random organ replacement needs. In addition, new players can experiment with surgery and no one is likely to be upset. Ideally, doctors can amputate all the usable parts in the early round time when little else is going on, to prepare for later and give them something to do.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: A couple of dead bodies are now in the morgue. The souls are long gone, so no one will care if you "borrowed" a part or two.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

zxaber committed:
Merge pull request #1 from tgstation/master

herp a derp this better not show up on the discord

zxaber committed:
DEAD

zxaber committed:
Update mapping_helpers.dm

QualityVan commented:
I don't dislike this, but couldn't they just hit up genetics for a monkeyman?

zxaber commented:
I suppose. I have no idea how hard or easy it is to make one. I just remember the few times I played medical being bored out of my mind and was hit by this idea earlier today.

MrPerson commented:
This is also pretty flavorful. Makes the station more lived-in.

Also more flavorful for the chef, if you get my drift.

Angustmeta commented:
Dilute one of the 50u mutadone pills in genetics in 10u water. Extinguisher 5 monkey cubes. Dropper 1u of the solution into the monkeys' eyes. = 5 monkeymen.

JJRcop commented:
Low random chance for husk or other strange damage?

anconfuzedrock commented:
This dude literally doesn't know how to make a fucking humanized monkey and makes this PR

ridiculous way to remove all possible effort required for any kind of biomatter, literal free chef meat, dumb as fuck

ATH1909 commented:
I think this is more of a flavor thing (like the dirt and bloodstains in the brig) than a "I can't seem to acquire any unneeded bodies as a doctor!" thing.

Tlaltecuhtli commented:
why 😂

SouDescolado commented:
About the MapHelper routine, how heavy is it on the server? With this same kind of code, you could make that all gray lockers get random items and not grab everything under it on roundstart. Could this be a method to increase performance and add more randomness to the game?

Menshin commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -205,3 +205,22 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+	for (var/i in a.contents)
+		if (istype(i, /obj/structure/bodycontainer/morgue))
+			trays += i
+	for (var/i = 1 to 2)
```
How about a `#define CORPSES_ON_ROUNDSTART` ?

ExcessiveUseOfCobblestone replied:
make it a var on the mapping_helper so you can manually edit it if your morgue is bigger/smaller than "normal"

Menshin replied:
^ better idea

zxaber replied:
Cobby's suggestion is done on the upcoming change.

Menshin commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -205,3 +205,22 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+		h.forceMove(j)
+		j.update_icon()
+	qdel(src)
+	return
```
This `return` is useless in Byond and can go.

ExcessiveUseOfCobblestone replied:
unneeded

Menshin replied:
This `return` is unneeded here and can go. 😉

zxaber replied:
Alright, gone in the upcoming change.

zamolxius commented:
this is great and "fluffy" if you know what i mean. now i can roleplay a necrophile medical doctor correctly.

kopoba commented:
Cool. I wish we could do something like "reviving surgery"(so body can be occupied free ghost) or "autopsy for points" on them.

anconfuzedrock commented:
Am I fucking crazy for thinking that this is completely excess fluff that does nothing but make morgue more annoying to sift through and less atmospheric knowing that not all of the bodies are actual dead players and give the cook way more free meat for 100% no downside? I will fucking die on this hill, this pr just makes the morgue *less* fluffy since now in a multiplayer game you can't think "wow, every one of these trays was a player who died." And the very very rare ridiculous circumstance of someone needing organs? All this does is make that circumstance mind-numbingly simple if you don't even need to make a fucking human monkey or get organs from other people, you can just pry them out of the free organ bin. ridiculous.

PKPenguin321 commented:
I think it's great flavor and the gameplay impacts are negligible until proven otherwise by actual gameplay

anconfuzedrock commented:
Thing is, zxaber isn't adding this for flavor. he's adding it for gameplay, which I think is ridiculous babying, and as for flavor, What if I, personally, think this is godawful flavor that detracts from the multiplayer and the roleplaying aspect of a multiplayer roleplaying game? Because I do. If some of the soapstone scribblings were just randomly generated, I wouldn't be interested in soapstones knowing one could be garbage not even written by a human being. If some of the library books were procedurally generated or ripped wikipedia articles, I wouldn't be as interested when I can't be sure if this book has an actual history or is just numbers no person has ever looked at. The bodycount of the morgue is the actual bodycount of people who died and were successfully taken care of by medbay, and every single body has at least some reason why it's there, even that husk with no soul in the corner. now that husk with no soul in the corner's history could be "it spawned there at roundstart" and that's just bland as all fuck.

Lobachevskiy commented:
This will affect the balance of cargo bounties. Currently there are things like get 3 livers, this just gives 2 of those for free every round, guaranteed.

What's the flavor anyway? Random NPCs had heart attacks? If there was some kind of moral issue at the very least, I'd agree, but we all know this will be simply treated as meat and organs. Hell, even the PR author says it's just "extra supplies".

At the first glance it looked like a great addition, but the more I think about, the less I like it.

zxaber commented:
I'm not like super committed to this PR, rock. If it's actually a bad idea and gets turned down, I'm not gonna lose sleep over it. You are right, the main purpose was gameplay. I just remember every round I've played as doctor (granted, hasn't been a bunch), where until someone managed to get hurt, I was just standing around and watching the vent fans spin. And so I thought "hey, what could we do to give doctors at least something to fill their time with before the shit hits the fan? Oh, what if there were one or two dead bodies in the morgue, not tied to players, that doctors could use for parts? Then the first ten to fifteen minutes could be spent doing surgery to remove the parts and stock up for when they're needed."

If that's the wrong way to go about it then ¯\\_(ツ)_/¯. Anyway, there's nothing wrong with being passionate about a change you think is wrong.

As far as bounties go, yeah, it'll affect them. I thought they were all more than three, but I might be wrong. If it's an issue, we _could_ just bump up the bounty amounts, but that feels like going the wrong way. I could also just drop the number of bodies to one, so that it's only one free part of whatever. Or, or additionally, I could remove random organs from the bodies, which might also at least be more interesting that two full bodies sitting there.

anconfuzedrock commented:
I get livid about everything. I think it isn't that I hate this because it's awful and will kill the game, but just that there's no advantage for all it does. Plenty of ways to get bodies already when genetics starts with so many monkey cubes, though maybe that isn't as common sense to everyone since genetics was my first job. I just think that all does is already easy enough, so this is pretty lazy and makes meat even more ridiculously easy for the chef than it already is for little good reason. Besides monkey cubes and suicides, medbay also has prosthetics and can eventually print better organs, so I just think this isn't worth it. Generally I just hate prs like this because I'm horrified it will somehow slippery slope into something nasty because I'm mentally unstable

habader commented:
red label ASAP

Grazyn commented:
Revenant buff

davethwave commented:
Just a reminder but revenants can only spawn with a large amount of crew dead so a few dead bodies at round start doesn't effect anything...

ExcessiveUseOfCobblestone commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -205,3 +205,22 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+			trays += i
+	for (var/i = 1 to 2)
+		var/obj/structure/bodycontainer/morgue/j = pick(trays)
+		var/mob/living/carbon/human/h = new /mob/living/carbon/human(j.loc, 1)
```
the morgue's `loc` is the turf so don't you want it inside the morgue instead?

zxaber replied:
All addressed.

zxaber committed:
updates

zxaber commented:
- Number of bodies is now set by the variable "bodycount".

- Cleaned up some of the code.

- Every organ for each body has a 20% chance of being missing.

Also, I'm running into an issue with the morgue trays. Despite me calling the update_icon() proc, they occasionally don't update their sprite to show the indicator that a body is inside. Not yet sure how to fix it.

Menshin commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -205,3 +205,24 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+		var/mob/living/carbon/human/h = new /mob/living/carbon/human(j, 1)
+		h.death()
+		for (var/part in h.internal_organs) //randomly remove organs from each body
+			if (rand(0,10) <=2)
```
```suggestion
			if (prob(20))
```
You've got a 3 chances out of 11 with you check.

zxaber replied:
oops

Menshin commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -205,3 +205,24 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+	icon_state = "deadbodyplacer"
+	var/bodycount = 2 //number of bodies to spawn
+
+/obj/effect/mapping_helpers/dead_body_placer/Initialize(mapload)
```
Should probably `LateInitialize()` so every trays in the area are instanciated before trying to put corpses in them.

zxaber replied:
Thanks, that did it.

zxaber committed:
more update

zxaber commented:
- Now late-loads to allow morgues time to exist

- Changed organ missing probability to 40%. I _think_ this means there's a 20% chance for both bodies to be missing a particular organ, if I remember how probability works.

- Added a catch to prevent runtiming if a mapper places the map helper outside an area with a morgue, and added a map log entry for when this happens.

ExcessiveUseOfCobblestone commented:
>     * Changed organ missing probability to 40%. I _think_ this means there's a 20% chance for both bodies to be missing a particular organ, if I remember how probability works.

It's higher than that since the prob is per organ and not per person. The probability NOTHING happens per person is <8% assuming 5 organs (60%^5), and the probability NOTHING happens to BOTH persons is <1% also assuming 5 organs (<8%^2). More organs = even lower chance.

zxaber commented:
Well, sure, but when talking about any particular organ, the other organs' dice rolls don't matter. If corpse A and B each have a 40% chance to be missing the liver, what is the probability that _both_ are missing the liver?

ExcessiveUseOfCobblestone commented:
low

Menshin commented:
> Well, sure, but when talking about any particular organ, the other organs' dice rolls don't matter. If corpse A and B each have a 40% chance to be missing the liver, what is the probability that both are missing the liver?

Making the reasonable assumption both rolls are statistically independent, that would be (40%)^2 = 16%.

zxaber commented:
I googled for a probability calculator to see what someone possibly smarter than I would say about it. This is from calculator.net, though I have no idea how trustworthy the site is.

![image](https://user-images.githubusercontent.com/37497534/57186036-e2bc9f80-6e8c-11e9-9c38-2bdda258d195.png)

So yeah, as Menshin said, a 16% chance that you won't get a particular organ from either body, which seems fine.

ExcessiveUseOfCobblestone commented:
whoops I misread ur post, it's just .4^2 as menshin said for a singular organ.

chesse20 commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -205,3 +205,28 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+	icon_state = "deadbodyplacer"
+	var/bodycount = 2 //number of bodies to spawn
+
+/obj/effect/mapping_helpers/dead_body_placer/LateInitialize()
```
code looks fine to me

chesse20 commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -205,3 +205,28 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+		var/obj/structure/bodycontainer/morgue/j = pick(trays)
+		var/mob/living/carbon/human/h = new /mob/living/carbon/human(j, 1)
+		h.death()
+		for (var/part in h.internal_organs) //randomly remove organs from each body
```
looks good

chesse20 commented on _maps/map_files/PubbyStation/PubbyStation.dmm:
```diff
@@ -89165,7 +89169,7 @@ bil
 bjT
 ble
-bmq
+aab
```
what does this do

Tlaltecuhtli replied:
map change

chesse20 commented on _maps/map_files/MetaStation/MetaStation.dmm:
```diff
@@ -109709,7 +109713,7 @@ ctA
 cCj
 cCj
 cDY
-cCj
```
what did this code do

Tlaltecuhtli replied:
map change

chesse20 commented on _maps/map_files/BoxStation/BoxStation.dmm:
```diff
@@ -4997,6 +4997,10 @@
 	},
 /turf/open/floor/plating,
 /area/ai_monitored/turret_protected/aisat_interior)
+"ajY" = (
```
looks good to me

chesse20 reviewed (approved):
looks great lets pr this

Jalleo commented:
This PR adds a third option. The chaplain doing two funerals at the same time.

anconfuzedrock commented:
is the "they miss an organ sometimes" think supposed to solve anything? people basically never need organs already, all this means is on the off chance someone needs one and they open up the morgue there's a slight chance they'll get a body with no organ and they'll have to get another and they'll be really frustrated about it. Plus again so many other corpses and with cyber organs, all this does is make the morgue a WORSE place to get organs. Kind of like if you gave chemists a bottle of oil roundstart so they wouldn't have to mix it but there was a 20% chance of that oil bottle actually being HIV, which they'll see when they try to use it to mix something and fail. I don't see the point in this pr, 2 free corpses with probably organs is already ridiculously excessive when the room literally next to the morgue can shit out like 10 catatonic bodies to use and you don't even need to know how to use a genetics console at all if you use the mutadone pills already found in genetics to grow humans from monkeys. This isn't helping doctors with any actual issue, this is putting 2 extra bodies with no story behind them in the morgue at best, confusing doctors into wasting time trying to harvest organs and giving chefs more FREE 100% OF THE TIME ROUNDSTART meat than they'll ever need so they never even have to consider trying to get more when they have plenty of options to do so, botany alone has several (And that is exactly what's going to happen, if this is supposed to help doctors it won't because 90% of the time the chef will grab both first and he won't have to feel bad because again doctors do not need these anyways.)

TLDR the organ solution basically does nothing and the chef is going to just take both bodies roundstart anyways, babying him instead of babying doctors who could get a humanized monkey in about 5 seconds from the room right next to the morgue.

ExcessiveUseOfCobblestone commented:
it's just fluff

ExcessiveUseOfCobblestone commented:
also chesse don't make me have to look at your code reviews only to find out they're crap. The questions were fine, the "good" ones are pretty useless.

zamolxius commented:
this is great, stop whining you little crybaby @anconfuzedrock it's not like the chef cant just get the geneticists' or xenobio spare monkeys, or just people who died and are ghosted, it's a nice touch

anconfuzedrock commented:
Yeah, its not like the chef can’t just get spare monkeys, which is the literal fucking flaw of this pr, you minge. Who do you think you are?
